tests/component/test_stream_readers.py

```diff
@@ -545,3 +550,598 @@ def test___power_binary_reader___read_many_sample_with_wrong_dtype___raises_erro
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    # 6363 port 2 has 8 lines, and DAQ will happily extend the data to the larger type.
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[2].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    return task
+
+
+@pytest.fixture
+def di_single_channel_port_uint32_task(
+    task: nidaqmx.Task, sim_6363_device: nidaqmx.system.Device
+) -> nidaqmx.Task:
+    # 6363 port 0 has 32 lines
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[0].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    return task
+
+
+@pytest.fixture
+def di_multi_channel_port_uint32_task(
+    task: nidaqmx.Task, sim_6363_device: nidaqmx.system.Device
+) -> nidaqmx.Task:
+    # 6363 port 0 has 32 lines
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[0].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    # 6363 port 1 has 8 lines, and DAQ will happily extend the data to the larger type.
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[1].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    # 6363 port 2 has 8 lines, and DAQ will happily extend the data to the larger type.
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[2].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    return task
+
+
+def _get_num_lines_in_task(task: nidaqmx.Task) -> int:
+    return sum([chan.di_num_lines for chan in task.channels])
+
+
+def _get_expected_digital_data_for_sample(num_lines: int, sample_number: int) -> int:
+    result = 0
+    # Simulated digital signals "count" from 0 in binary within each group of 8 lines.
+    for _ in range((num_lines + 7) // 8):
+        result = (result << 8) | sample_number
+
+    line_mask = (2**num_lines) - 1
+    return result & line_mask
+
+
+def _get_expected_digital_data(num_lines: int, num_samples: int) -> List[int]:
+    return [
+        _get_expected_digital_data_for_sample(num_lines, sample_number)
+        for sample_number in range(num_samples)
+    ]
+
+
+def _get_expected_digital_port_data_port_major(
+    task: nidaqmx.Task, num_samples: int
+) -> List[List[int]]:
+    return [_get_expected_digital_data(chan.di_num_lines, num_samples) for chan in task.channels]
+
+
+def _get_expected_digital_port_data_sample_major(
+    task: nidaqmx.Task, num_samples: int
+) -> List[List[int]]:
+    result = _get_expected_digital_port_data_port_major(task, num_samples)
+    return numpy.transpose(result).tolist()
+
+
+def _bool_array_to_int(bool_array: numpy.typing.NDArray[numpy.bool_]) -> int:
+    result = 0
+    # Simulated data is little-endian
+    for bit in reversed(bool_array):
+        result = (result << 1) | bit
+    return result
+
+
+def _read_and_copy(
+    read_func: Callable[[numpy.typing.NDArray], None], array: numpy.typing.NDArray
+) -> numpy.ndarray:
```

Comment:
FYI, the type checker doesn't know all three of these ndarrays have the same dtype. Here's how to tell it that (I think):

```
_DType = TypeVar("_DType", bound=numpy.typing.DTypeLike)

def _read_and_copy(
    read_func: Callable[[numpy.typing.NDArray[_DType]], None], array: numpy.typing.NDArray[_DType]
) -> numpy.typing.NDArray[_DType]:
```

Python 3.12 makes this less confusing by removing the separate TypeVar declaration, but we can't use this syntax until 3.11 is EOLed:
```
def _read_and_copy[DType: numpy.typing.DTypeLike](
    read_func: Callable[[numpy.typing.NDArray[DType]], None], array: numpy.typing.NDArray[DType]
) -> numpy.typing.NDArray[DType]:
```

Reply:
Neat! That matches what I see in the documentation, and it seems to work.